go/teams/oracle.go

```diff
@@ -0,0 +1,88 @@
+package teams
+
+import (
+	"github.com/keybase/client/go/libkb"
+	"github.com/keybase/client/go/protocol/keybase1"
+	"golang.org/x/crypto/nacl/box"
+)
+
+func TryDecryptWithTeamKey(mctx libkb.MetaContext, arg keybase1.TryDecryptWithTeamKeyArg) (ret []byte, err error) {
+	loadArg := keybase1.LoadTeamArg{
+		ID:          arg.TeamID,
+		Public:      arg.TeamID.IsPublic(),
+		ForceRepoll: false,
+		Refreshers: keybase1.TeamRefreshers{
+			NeedKeyGeneration: arg.MinGeneration,
+		},
+	}
+	team, err := Load(mctx.Ctx(), mctx.G(), loadArg)
+	if err != nil {
+		return nil, err
+	}
+
+	mctx.CDebugf("Loaded team %q, max key generation is %d", team.ID, team.Generation())
+
+	tryKeys := func(min keybase1.PerTeamKeyGeneration) (ret []byte, found bool, err error) {
+		if min == 0 {
+			// per team keys start from generation 1.
+			min = 1
+		}
+		for gen := min; gen <= team.Generation(); gen++ {
+			key, err := team.encryptionKeyAtGen(gen)
+			if err != nil {
+				mctx.CDebugf("Failed to get key gen %d: %v", gen, err)
+				switch err.(type) {
+				case libkb.NotFoundError:
+					continue
+				default:
+					return nil, false, err
+				}
+			}
+
+			mctx.CDebugf("Trying to unbox with key gen %d", gen)
+			decryptedData, ok := box.Open(nil, arg.EncryptedData[:], (*[24]byte)(&arg.Nonce),
+				(*[32]byte)(&arg.PeersPublicKey), (*[32]byte)(key.Private))
+			if !ok {
+				continue
+			}
+
```

Comment:
done, thanks.

if you don't mind, I have another idea how to refactor this function to make it still do the "try from cache first, then repoll if didn't find key" routine, but be a bit less messy. I could try this in a bit but it will require a re-review

Reply:
sure thing